test/test_nn.py

```diff
@@ -11494,6 +11494,50 @@ def __init__(self):
                 self.assertEqual(p.grad.to(devices[0]), pe.grad)
 
 
+class TestLazyModules(TestCase):
+    def test_lazy_module(self):
+        module = torch.nn.Module()
+        module.register_parameter('test_param', torch.nn.parameter._UninitializedParameter())
+
+        with self.assertRaisesRegex(ValueError, 'uninitialized'):
+            list(module.parameters())
+
+        with self.assertRaisesRegex(ValueError, 'uninitialized'):
+            module._apply(lambda x: x)
+
+        state_dict = module.state_dict()
+        self.assertIsInstance(state_dict['test_param'], nn.parameter._UninitializedParameter)
+        new_module = torch.nn.Module()
+        # Uninitialized paramaters in a state dict are ignored when the module has a valid one.
+        new_module.register_parameter('test_param', nn.Parameter(torch.ones(5, 5)))
+        new_module.load_state_dict(state_dict) 
```

Comment:
I am not sure about this behavior. That means that random values might remain in the model while the user expect to have overwritten all of them.
It should not happen often in practice but I would argue we might want to raise an error here.

Reply:
Yes, this was a doubt that I have in the beginning, I completely agree to raise an error here. Thanks!

Reply:
now it raises `ValueError` in this case and initializes a full-fledged parameter in the other way around.